### PR TITLE
Don't quantize against sampler decks

### PR DESCRIPTION
**Pull request by ywwg**
This should fix the issue reported by @Holzhaus where decks were not quantizing correctly

### CODE REVIEW

**Review by daschuer — CHANGES_REQUESTED** [review]
The issue not only happens with samplers. 
Other stopped decks are effected as well. 
I think the current deck must only sync to stopped decks if it is not already playing.

**Comment by daschuer** [comment]
I think it needs a little more love. 

EngineControl::pickSyncTarget()
always calls pickNonSyncSyncTarget() 

But i think using a target independent from the sync state should be the exception. It should first return the current sync master if there is any, that the first synced deck and than the first non synced Deck. In addition it should always return a playing deck if there is any. 

Most caller of pickSyncTarget() fall back to itself if it is playing and no playing sync target is returned. 
Not sure if we can move this logic into the function as well. Maybe we should keep it extern to not break anything. 

pickNonSyncSyncTarget() is also called from BaseTrackPlayerImpl::slotCloneDeck() in case no Deck to clone from is explicitly selected. I think this can also make use from coling from the mastr deck instead of from any other deck.

**Comment by daschuer** [comment]
since it turns out the issue was my dogfood here is a PR: 
https://github.com/ywwg/mixxx/pull/7

However I think my comment above still applies. This can be done here or in a new PR. 

I don't mind if we should quantize to a single playing sampler.

**Comment by ywwg** [comment]
picksynctarget was only supposed to be called when sync is not enabled. It looks like its use has drifted because of poor naming

**Comment by ywwg** [comment]
ok updated

**Comment by ywwg** [comment]
> coling

not sure what this word is supposed to be.


There's a lot of redundancy between pickMaster and pickNonSyncSyncTarget. hm.

**Comment by daschuer** [comment]
Cloning. 

I am not sure if this is really happening, but when cloning without explicit name a deck, it can also default to a playing master.

**Comment by ywwg** [comment]
hmmm. the trouble here is that picksyncmaster wants to return an enginebuffer, and InternalClock has no enginebuffer.  So it has no beats, or beat distance, etc, so it can't be used to calculate divsec in getBeatMatchPosition.  I think we need to move the calculation of otherDivSec into Syncable itself so there can be different algorithms for decks vs internal clock

**Comment by ywwg** [comment]
yeah I think that's doable!  I'll poke at it

**Comment by Holzhaus** [comment]
What's the status of this PR? Ready to review or are you still working on it? In the latter case, please mark it as draft.

**Comment by ywwg** [comment]
PR updated and ready for review

**Comment by daschuer** [comment]
unfortunately 
EngineSyncTest.UserTweakPreservedInSeek 
Fails on AppVeyor

a rounding issue:
 
```
..\src\test\enginesynctest.cpp(1659): error: Expected equality of these values:
  0.025154950869236584
  ControlObject::get(ConfigKey(m_sGroup1, "beat_distance"))
    Which is: 0.025154950869236758
..\src\test\enginesynctest.cpp(1660): error: Expected equality of these values:
  0.0023582766439909299
  ControlObject::get(ConfigKey(m_sGroup1, "playposition"))
    Which is: 0.0023582766439909464
in ~EngineMaster()
[  FAILED  ] EngineSyncTest.UserTweakPreservedInSeek (920 ms)
```

**Comment by ywwg** [comment]
I was away for the weekend.  I'll fix the rounding issue.  Can a couple people try this patch manually and see if it works for them?

**Review by Holzhaus — COMMENTED** [review]
Unfortunately this does not fix the quantize issue for me.
I made a small video that demonstrates the issue: https://www.youtube.com/watch?v=Q1jjkuN9QTs

With two decks and sync/quantize enabled it works, but if I set explicit master on one deck (blinking sync button in the video) it stops working.

**Comment by ywwg** [comment]
explicit master is out of scope for this PR, I'm fixing that in a different PR

**Comment by Holzhaus** [comment]
> explicit master is out of scope for this PR, I'm fixing that in a different PR

I didn't realize that, sorry for the noise.

**Comment by ywwg** [comment]
no worries! the explicit master issues are more fundamental and complex so they require more work

File: src/engine/controls/bpmcontrol.cpp [hunk 1]
```diff
@@ -763,21 +771,15 @@ double BpmControl::getBeatMatchPosition(
...
     if (playing) {
-        // "this" track is playing, or just starting
-        // only match phase if the sync target is playing as well
-        if (pOtherEngineBuffer->getSpeed() == 0.0) {
-            return dThisPosition;
+        if (!pOtherEngineBuffer || pOtherEngineBuffer->getSpeed() == 0.0) {
+            // "this" track is playing, or just starting
+            // only match phase if the sync target is playing as well
+            // else use the previous phase of "this" track before the seek
+            pOtherEngineBuffer = getEngineBuffer();
```
Comment (daschuer) [inline src/engine/controls/bpmcontrol.cpp:779]: This also applies for explicit master. 

So please remove 
```
    if (getSyncMode() == SYNC_MASTER_EXPLICIT) {
        return dThisPosition;
    }
```
above.

File: src/engine/sync/enginesync.cpp [hunk 2]
```diff
@@ -429,31 +429,46 @@ void EngineSync::deactivateSync(Syncable* pSyncable) {
...
 
-EngineChannel* EngineSync::pickNonSyncSyncTarget(EngineChannel* pDontPick) const {
-    EngineChannel* pFirstNonplayingDeck = nullptr;
+Syncable* EngineSync::pickNonSyncSyncTarget(EngineChannel* pDontPick) const {
+    // First choice: the sync master, if it's a deck
+    if (m_pMasterSyncable &&
+            m_pMasterSyncable->getChannel() &&
+            m_pMasterSyncable->getChannel() != pDontPick) {
+        return m_pMasterSyncable;
+    }
```
Comment (daschuer) [inline src/engine/sync/enginesync.cpp:438]: An explicit master should be never synced to anything else. So we should return nullptr if pDontPick is an expicite master.

**Review by daschuer — CHANGES_REQUESTED** [review]

**Comment by daschuer** [comment]
Ping!

**Comment by ywwg** [comment]
back on this, sorry

**Comment by ywwg** [comment]
agh sorry wrong PR

File: src/engine/controls/bpmcontrol.cpp (hunk 1 above)
Comment (ywwg) [reply src/engine/controls/bpmcontrol.cpp:779]: done

File: src/engine/sync/enginesync.cpp (hunk 2 above)
Comment (ywwg) [reply src/engine/sync/enginesync.cpp:438]: it's not the job of this function to decide who should call it.  Instead, the calling function should check to see if it's an explicit master and shouldn't be looking for a sync source in the first place.  (Also, pickSyncTarget is also used by KeyControl, so this isn't only used for bpm matching).

Fixed by changing the call points of pickSyncTarget in bpmcontrol.

**Comment by ywwg** [comment]
rebased onto 2.3

**Comment by ywwg** [comment]
appveyor passing now

**Review by daschuer — APPROVED** [review]

**Comment by daschuer** [comment]
Thank you, I did a brief test and everything works fine. 
LGTM

**Pull request merged**